localstack/services/moto.py

```diff
@@ -0,0 +1,249 @@
+    provider, and wrapping each method with a fallthrough method that calls ``reqest`` if the original provider
+    raises a ``NotImplementedError``.
+
+    :param provider: the ASF provider
+    :return: a modified DispatchTable
+    """
+    table = create_dispatch_table(provider)
+
+    for op, fn in table.items():
+        table[op] = _wrap_with_fallthrough(fn)
+
+    return table
+
+
+def _wrap_with_fallthrough(handler: ServiceRequestHandler) -> ServiceRequestHandler:
+    def _call(context, req) -> ServiceResponse:
+        try:
+            # handler will typically be an ASF provider method, and in case it hasn't been implemented, we try to
+            # fall through to moto
+            return handler(context, req)
+        except NotImplementedError:
+            return proxy_moto(context)
+
+    return _call
+
+
+def dispatch_to_moto(context: RequestContext) -> MotoResponse:
+    """
+    Internal method to dispatch the request to moto without changing moto's dispatcher output.
+    :param context: the request context
+    :return: the response from moto
+    """
+    service = context.service
+    request = context.request
+
+    # hack to avoid call to request.form (see moto's BaseResponse.dispatch)
+    setattr(request, "body", request.data)
+
+    # this is where we skip the HTTP roundtrip between the moto server and the boto client
+    dispatch = get_dispatcher(service.service_name, request.path)
+
+    return dispatch(request, request.url, request.headers)
+
+
+def get_dispatcher(service: str, path: str) -> MotoDispatcher:
+    url_map = get_moto_routing_table(service)
+
+    if len(url_map._rules) == 1:
+        # in most cases, there will only be one dispatch method in the list of urls, so no need to do matching
+        rule = next(url_map.iter_rules())
+        return rule.endpoint
+
+    matcher = url_map.bind(config.LOCALSTACK_HOSTNAME)
+    endpoint, _ = matcher.match(path_info=path)
+    return endpoint
+
+
+@lru_cache()
+def get_moto_routing_table(service: str) -> Map:
+    """Cached version of load_moto_routing_table."""
+    return load_moto_routing_table(service)
+
+
+def load_moto_routing_table(service: str) -> Map:
+    """
+    Creates from moto service url_paths a werkzeug URL rule map that can be used to locate moto methods to dispatch
+    requests to.
+
+    :param service: the service to get the map for.
+    :return: a new Map object
+    """
+    # code from moto.server.create_backend_app
+    backend_dict: BackendDict = get_moto_backend(service)
+    if "us-east-1" in backend_dict:
+        backend = backend_dict["us-east-1"]
+    else:
+        backend = backend_dict["global"]
+
+    url_map = Map()
+    url_map.converters["regex"] = RegexConverter
+
+    for url_path, handler in backend.flask_paths.items():
+        # endpoints are annotated as string in werkzeug, but don't have to be
+        url_map.add(Rule(url_path, endpoint=handler))
+
+    return url_map
+
+
+def create_aws_request_context(
```

Comment:
yes! currently it is mainly used for testing. it can be useful in situations where you want to modify the AWS service request dictionary as it is passed to the ASF, and create a new request from that, rather than the original HTTP request.